### PR TITLE
[WIP] modify mixed result, throw if view file is null

### PR DESCRIPTION
**Pull request by kamarton**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | -

### CODE REVIEW

File: src/WebView.php [hunk 1]
```diff
@@ -138,6 +138,9 @@ public function head(): void
      */
     public function beginBody(): void
     {
+        if ($this->getViewFile() === null) {
+            throw new \LogicException('View file cannot be null.');
```
Comment (xepozz) [inline src/WebView.php]: What do you think about to create `ViewFileNotFoundException`?

File: tests/ViewTest.php [hunk 2]
```diff
@@ -142,6 +142,28 @@ private function createFileStructure(array $items, string $baseDirectory = null)
...
+
+    public function testViewFileNullEndPage(): void
+    {
+        $config = require Builder::path('tests');
+
+        $container = new Container($config);
+        $view = $container->get(View::class);
+
+        $this->expectException(\LogicException::class);
+        $view->endPage();
```
Comment (xepozz) [inline tests/ViewTest.php]: I think `endPage()` should fall down if `beginPage()` was not called before.
So this test case is wrong.

File: tests/ViewTest.php [hunk 3]
```diff
@@ -142,6 +142,28 @@ private function createFileStructure(array $items, string $baseDirectory = null)
...
+
+        $this->expectException(\LogicException::class);
+        $view->beginPage();
+    }
+
+    public function testViewFileNullEndPage(): void
+    {
+        $config = require Builder::path('tests');
+
+        $container = new Container($config);
```
Comment (xepozz) [inline tests/ViewTest.php]: Can you extract this to new method `createContainer()`/`getContainer()`?

File: src/View.php [hunk 4]
```diff
@@ -570,6 +570,9 @@ public function popDynamicContent(): void
      */
     public function beginPage(): void
     {
+        if ($this->getViewFile() === null) {
+            throw new \LogicException('View file cannot be null.');
```
Comment (samdark) [inline src/View.php]: We're not setting the file ourselves so the exception is the result of calling this method when no file is being rendered, right?

How about `\RuntimeException('beginPage() should be called in views only')`?

File: src/View.php [hunk 5]
```diff
@@ -583,6 +586,9 @@ public function beginPage(): void
      */
     public function endPage(): void
     {
+        if ($this->getViewFile() === null) {
```
Comment (samdark) [inline src/View.php]: How about `\RuntimeException('endPage() should be called in views only')`?

File: src/WebView.php (hunk 1 above)
Comment (samdark) [reply src/WebView.php]: That's not the case of view file not found... that's something like  `\RuntimeException('beginBody() should be called in views only')`...

File: src/WebView.php [hunk 6]
```diff
@@ -147,6 +150,9 @@ public function beginBody(): void
      */
     public function endBody(): void
     {
+        if ($this->getViewFile() === null) {
```
Comment (samdark) [inline src/WebView.php]: How about `\RuntimeException('endBody() should be called in views only')`?

File: src/WebView.php [hunk 7]
```diff
@@ -160,6 +166,9 @@ public function endBody(): void
      */
     public function endPage($ajaxMode = false): void
     {
+        if ($this->getViewFile() === null) {
```
Comment (samdark) [inline src/WebView.php]: How about `\RuntimeException('endPage() should be called in views only')`?

File: tests/ViewTest.php (hunk 2 above)
Comment (samdark) [reply tests/ViewTest.php]: In this case we should test for exception message string.

File: src/View.php (hunk 4 above)
Comment (kamarton) [reply src/View.php]: How about `\LogicException('beginPage  should be called in views context only')`?

`\LogicException`:
> Exception that represents error in the program logic. **This kind of exception should lead directly to a fix in your code.**

File: src/WebView.php (hunk 6 above)
Comment (kamarton) [reply src/WebView.php]: Are such cases possible? If not, should be warned to call `beginPage()` before `endPage()`. In this case, `getViewFile ()` cannot be null and unnecessary to check, but can check `if (BeginPage not called) {throw ...}`.

```php
$this->beginPage();
$this->beginPage();   // nested begin page
...
$this->endPage();
$this->endPage();
```

or

```
$this->endPage(); // call without beginPage
```

File: src/WebView.php (hunk 7 above)
Comment (kamarton) [reply src/WebView.php]: Are such cases possible? If not, should be warned to call `beginPage()` before `endPage()`. In this case, `getViewFile ()` cannot be null and unnecessary to check, but can check `if (BeginPage not called) {throw ...}`.

```php
$this->beginPage();
$this->beginPage();   // nested begin page
...
$this->endPage();
$this->endPage();
```

or

```
$this->endPage(); // call without beginPage
```

File: src/View.php (hunk 4 above)
Comment (samdark) [reply src/View.php]: Yes, that's fine.

**Comment by kamarton** [comment]
I close it, because the operation of the expected operation is not clear, see more: #78

**Pull request closed without merging**